### ISSUE TITLE
LoopingCall.withCount countCallable called with 0

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7836]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| d.vinella reported|
|-|-|
|Trac ID|trac#7836|
|Type|defect|
|Created|2015-03-26 09:48:03Z|
|Branch|https://github.com/twisted/twisted/tree/count-7836|

Sometimes using a LoopingCall.withCount with a very small interval (0.02 sec) the countCallable is called with 0 as argument and it looks like the now - _realLastTime is smaller than the looping interval. The docstring for LoopingCall.withCount says nothing about the possibility of a calls arg == 0
NTP is not enabled during the execution; happens on linux and osx with medium/heavy load; [mailing list discussion](https://twistedmatrix.com/pipermail/twisted-python/2015-March/029316.html)

Code to reproduce:

```
import time
 
from twisted.internet import reactor
from twisted.internet.task import LoopingCall
from twisted.python import log
from twisted.python.logfile import LogFile
 
log.startLogging(LogFile('lc.log', '.'))
 
 
def count_callable(calls):
    log.msg(format='***************** %(t)f calls:%(c)d', t=time.time(), c=calls)
    if calls != 1:
        log.msg('******************************************* anomaly')
 
 
def debug_lc(lc):
    log.msg(format="_expectNextCallAt %(x)f", x=lc._expectNextCallAt)
    log.msg(format=" _realLastTime %(x)f", x=lc._realLastTime)
 
 
lc = LoopingCall.withCount(count_callable)
lc.start(0.02)
 
lc_debug = LoopingCall(debug_lc, lc)
lc_debug.start(0.02)
 
reactor.run() 
```

log

```
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.325579 calls:1
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.345534
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.325566
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.345573 calls:1
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.365534
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.345560
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.365558 calls:0
2015-03-23 17:36:30+0100 [-] ******************************************* anomaly
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.385534
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.365550
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.385583 calls:2
2015-03-23 17:36:30+0100 [-] ******************************************* anomaly
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.405534
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.385568
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.405578 calls:1
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.425534
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.405564
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.425597 calls:1
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.445534
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.425575
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.445692 calls:1
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.465533
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.445668
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.465584 calls:1
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.485533
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.465571
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.485560 calls:0
2015-03-23 17:36:30+0100 [-] ******************************************* anomaly
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.505533
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.485552
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.505584 calls:2
2015-03-23 17:36:30+0100 [-] ******************************************* anomaly
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.525533
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.505566
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.525602 calls:1
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.545533
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.525585
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.545703 calls:1
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.565533
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.545681
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.565583 calls:1
2015-03-23 17:36:30+0100 [-] _expectNextCallAt 1427128590.585533
2015-03-23 17:36:30+0100 [-]     _realLastTime 1427128590.565570
2015-03-23 17:36:30+0100 [-] ***************** 1427128590.585577 calls:1
```

<details><summary>Searchable metadata</summary>

```
trac-id__7836 7836
type__defect defect
reporter__d_vinella d.vinella
priority__normal normal
milestone__ 
branch__branches_count_7836 branches/count-7836
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1427363283703385 1427363283703385
changetime__1431174230178559 1431174230178559
version__None None
owner__glyph glyph
cc__d.vinella
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Thanks for filing this bug.

Originally (as per the mailing list thread, thanks for linking that) I felt that this issue was due to time going backwards.  But that seems to be pretty clearly not happening on systems where this is occurring, so at the very least it's not _necessary_ for it to occur ;).  If it's never going backwards, then we should be protected from short timeouts by the code in `twisted.internet.base.runUntilCurrent` which prevents the `callLater` from ever being invoked.

One problem with the log attached here is that it's calling `time.time()` itself, not looking at the `now` discovered by the `counter` nested function inside `LoopingCall.withCount`.  This means it's not reporting all the data that the actual `LoopingCall` algorithm is receiving (and the value of `time.time` at that particular point in program execution isn't super interesting).

Now I think that this has something to do with floating point inaccuracy, and I will investigate in that direction.

The way that `LoopingCall` determines the value for `withCount` is by bucketing both the current and previous invocation timestamps into "intervals".  It does this with floating-point division.  So if the numbers are close to an interesting boundary, they might both end up in the same interval even if they really "should" be in different ones.

I'm going to play with some numbers and see if I can come up with a unit test.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Yup, it's definitely floating-point math.  Just advance a clock attached to a `LoopingCall` of duration `0.1` ten times and you will see one count=0 call.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

(In [44332]) Branching to count-7836.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

So, the branch now contains a fix, although it needs a little more polish to be an acceptable change.  d.vinella, can you try this with your example code and tell me if it fixes the problem and doesn't introduce any others? :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|d.vinella commented|
|-|-|

Yep, I'll check it asap!

ps: sorry for the delay but I've not received any mail notification for the updates on this ticket, even if I opened it. Did I miss some preference in Trac, or do I have to add myself to cc when I open tickets?

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|d.vinella commented|
|-|-|

I can confirm that my use case is fixed with this patch, and it seems not to break anything.
Thank you very much for your support glyph, I hope to see this merged in trunk soon!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

Finally managed to update the test docstring and add a topfile; spinning the buildbots now and hopefully a reviewer can get to it soon.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Oh, oops, I forgot about `assertGreaterEqual`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

Fix for that is in as well, builders running.  The twistedchecker issue I see is spurious; possibly because this branch is too out of date with current trunk.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

Builds are green :)

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_12"></a>|@moshez set owner to @glyph|
|-|-|

Nit: docstring of test is unclear. Maybe s/When/A/?
In the test, how was the number of zeroes determined? would 0.05 be equally useful
The assertTrue should be assertGreaterEqual
Does the assertTrue belong in this test? It's just testing clock.advance is correct. This seems like a useful test
to add to the task.Clock tests
The howLong function is unnecessary -- writing the code inline is just as many indentation levels.
Also, since this is already subtle stuff, I'd avoid using the "x if y else z" thing and break it out to multiple lines.
This also applies to the triply-nested expression -- please break into separate lines, and name intermediate results.
In addition, comments for why this scheduling subtlety is necessary inline would be useful.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

Thanks for the review, moshez.  Welcome back :-).

A few meta-review comments, not all of which are actually covered in ReviewProcess, so I really need to update that to be a lot clearer.

  1. You didn't number your feedback; the convention is to use a numbered list (like this one) so it's easy to refer to individual review points.
  2. A bit self-serving to say this for myself, but you didn't say "thanks" - we always try to start code reviews off by thanking the contributor (even if the code is totally unusable, thanking them for their time and effort).
  3. You didn't actually remove the "review" keyword so this isn't a review yet :).
  4. You didn't separate this feedback into "what needs doing" versus "some things I noticed" sections, and you didn't make it clear what the next step should be (as described in https://twistedmatrix.com/trac/wiki/ReviewProcess#Howtobeagoodreviewer )

Replying to [moshez](#note_12):
> Nit: docstring of test is unclear. Maybe s/When/A/?

Yeah, that is just a grammatical error, I need to fix it.

> In the test, how was the number of zeroes determined? would 0.05 be equally useful

It's a value that's a little bit greater than `1.0 - (0.1+ 0.1+ 0.1+ 0.1+ 0.1+ 0.1+ 0.1+ 0.1+ 0.1+ 0.1)`.  `0.05` is a lot greater :).  It would probably be clearer to write it as `1.0e-something` rather than typing out all the zeroes.

> The assertTrue should be assertGreaterEqual

No, it shouldn't.  It was originally, but we support Python 2.6, and that method isn't available there, so some of the builders fail.

> Does the assertTrue belong in this test? It's just testing clock.advance is correct. This seems like a useful test
> to add to the task.Clock tests

It's a sanity check to help maintainers who might want to change the constants, to ensure that it's being advanced by an appropriate amount.  These sorts of checks aren't uncommon.

> The howLong function is unnecessary -- writing the code inline is just as many indentation levels.

Yes, but by making it a function rather than a variable assignment, I can use it as an expression with a return value, which lets me use the two return statements to indicate two separate values without having to be careful that my "if" tree is assigning the result value in all cases.  When you say it's "unnecessary" that makes it sound like function definitions are a scarce resource we are trying to conserve :).

> Also, since this is already subtle stuff, I'd avoid using the "x if y else z" thing and break it out to multiple lines.

That's fair; I think that code was the result of attempting to eliminate the multiple assignments to a result value and convert it into an expression, before giving up and putting it into a nested function.  I'll just commit to the nested-function strategy and make it more straightforward :).

> This also applies to the triply-nested expression -- please break into separate lines, and name intermediate results.

I assume you mean the one beginning `self.interval -`...?

> In addition, comments for why this scheduling subtlety is necessary inline would be useful.

Sure.  It's ... surprisingly hard to explain.

Please remove the 'review' keyword so that you get appropriate credit for the review on https://twistedmatrix.com/highscores/ - thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @moshez|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_15"></a>|@moshez set owner to @glyph|
|-|-|

Sorry :)

To clarify -- yes, I mean the self.interval calculation would be more readable broken up, and with intermediate results given meaningful (as can be) names.

Also, thank you for working on this -- this is one annoying bug :)

Removing review kw, assigning back to you -- hopefully I got it correct enough this time!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

Replying to [moshez](#note_15):
> Sorry :)
> 
> To clarify -- yes, I mean the self.interval calculation would be more readable broken up, and with intermediate results given meaningful (as can be) names.
> 
> Also, thank you for working on this -- this is one annoying bug :)
> 
> Removing review kw, assigning back to you -- hopefully I got it correct enough this time!

Pretty good :).  The one final detail is that you should always say "please address and re-submit" or "please address and land" if you are reviewing for another committer.

One is obliged to be conservative and assume it's "re-submit" if not otherwise specified, so I'll get to work on responding to this feedback now…

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

(In [44687]) clarify the grammar in the test docstring

refs [#7836](https://github.com/twisted/twisted/issues/7836)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

(In [44688]) compute the epsilon value exactly

rather than relying on "a very small value".

this should clarify why this particular value was selected.

refs [#7836](https://github.com/twisted/twisted/issues/7836)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

(In [44689]) explain in much more careful detail what is being tested

refs [#7836](https://github.com/twisted/twisted/issues/7836)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

(In [44690]) this code is highly subtle

so explain its nuances in detail in comments

refs [#7836](https://github.com/twisted/twisted/issues/7836)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph removed owner|
|-|-|

Resubmitting; I think we should be good to go now.

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_22"></a>|@moshez set owner to @glyph|
|-|-|

Hi glyph, thank you for working on this.

This seems good to go -- please land this. All issues I pointed out have been addressed!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_23"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44692]) Merge count-7836: Fix an issue with twisted.internet.task.LoopingCall.withCount where sometimes the passed callable would be invoked with "0" when we got close to tricky floating point boundary conditions.

Author: glyph
Reviewer: moshez
Fixes: [#7836](https://github.com/twisted/twisted/issues/7836)
